plugins/rebalance/rest.go

```diff
@@ -0,0 +1,323 @@
+package rebalance
+
+import (
+	"net/http"
+
+	"github.com/gluster/glusterd2/glusterd2/gdctx"
+	restutils "github.com/gluster/glusterd2/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+        "github.com/gluster/glusterd2/pkg/errors"
+	rebalanceapi "github.com/gluster/glusterd2/plugins/rebalance/api"
+
+	"github.com/gorilla/mux"
+	"github.com/pborman/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+func createRebalanceInfo(volname string, req *rebalanceapi.StartReq) *rebalanceapi.RebalInfo {
+	return &rebalanceapi.RebalInfo{
+		Volname:     volname,
+		RebalanceID: uuid.NewRandom(),
+                Status:      rebalanceapi.Started,
+		Cmd:         checkCmd(req),
+		CommitHash:  setCommitHash(),
+		RebalStats:       rebalanceapi.RebalNodeStatus{},
+	}
+}
+
+func rebalanceStartHandler(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	// collect inputs from url
+	volname := mux.Vars(r)["volname"]
+
+	var req rebalanceapi.StartReq
+
+//  TODO : How do I pass these options?
+//  Unmarshal Request so to handle fix-layout and start force
+
+//      if err := restutils.UnmarshalRequest(r, &req); err != nil {
+//              restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrJSONParsingFailed)
+//              return
+//	}
+
+	vol, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+		return
+	}
+
+	if vol.State != volume.VolStarted {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrVolNotStarted)
+		return
+	}
+
+	if vol.DistCount == 1 {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrVolNotDistribute)
+		return
+	}
+
+	// Check for remove- brick pending
+	//TODO
+
+	// A simple transaction to start rebalance on all nodes
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+
+	lock, unlock, err := transaction.CreateLockSteps(volname)
```

Comment:
There has been some changes regarding when we do transaction locking and unlocking.
Refer #510 for why, and #634 and #635 for how.
In short, we need to lock, unlock outside of the transaction, before we do the first GetVolume.
Create the lock/unlock funcs, lock and defer unlock.
```go
lock, unlock, err := transaction.CreateLockFuncs(volname)
lock(ctx)
defer unlock(ctx)
```
@prashanthpai began doing changes

Reply:
When does the unlock go through?

Reply:
When the handler function ends. The `defer` keyword adds the given function to be run after the current function calls return. You can defer as many functions as you want, and they will all get called, in LIFO order, at the end.